doc/design.md

```diff
@@ -0,0 +1,93 @@
+# Kubernetes MAC pool address allocation
+
+## Motivation
+
+As of date, there is no way to ensure MAC address collisions do not occur within a Kubernetes cluster. Many CNI plugins allocate MAC addresses randomly, with no assurance of collision prevention.
+
+Even if a CNI plugin ensures it never creates collisions, a cluster operating with multiple CNI plugins or sharing a broadcast domain with another cluster will remain collision-prone.
+
+## Limitations
+CNI plugins that do not implement CNI's MAC address specification will not be supported. The cluster must have multus installed and configured.
+
+## Goals
+
+This document proposes a collision-free allocation of MAC addresses within a single cluster.
+
+### Allocation
+This specification will describe an intermediate step prior to the initialization of Pods within a kubernetes cluster. During this step, a Pod's template will be read, processed, and modified to specify a MAC address.
+
+It is also to be noted that besides Pods, the proposed behavior will be generic enough to support future entities that may require MAC address allocation. For convenience sake, I will only refer to Pods throughout this document.
+
+### Collision prevention
+This specification will describe how the allocation process (as described in 3.1) will ensure no collisions may occur.
+
+## Design
+Both the act of modifying a Pod's template and deriving available addresses will be done by a new designated admission controller. The admission controller will operate within a predefined address range. A user will be able to pass his own desired range, or let a default range take place.
+
+### Allocation
+In order to control the allocation of MAC addresses for network devices, the admission controller will intercept the initialization of a Pod and observe its NetworkAttachmentDefinition's "mac" field.
+
+There are 2 possibilities for the "mac" field:
+
+It is unpopulated. In such case, the CNI assigns a MAC address.
+
+A static MAC address is present. In such case, the CNI assigns the specified address.
```

Comment:
Mention whether we check that desired address is already allocated

Reply:
done